### PR TITLE
Qingstor

### PR DESCRIPTION
**Pull request by Kael9527**
Hi, QingStor is an object storage service provided by QingClound, Similar to S3. And I start working on QingStor support.
        
Thanks for taking look first!

### CODE REVIEW

**Comment by ncw** [comment]
Hey, that looks amazing!

Some questions for you

  * Do you know if can I get a free account for testing (I'm based in the UK)?
  * I see it is passing most of the tests - well done!  What about the remaining ones? I'm willing to bend the tests to make Qingstor pass, but I'd like all the remotes to be running all the tests.
  * have you tried the tests in `fs` with `go test -v -remote TestQS:`?
  * It says in the docs (translated by google translate) that it is AWS S3 compatible - does the S3 remote not work properly with Qingstor?

And an apology - I'm just about to change the `List` interface, splitting it into a simple `List` which just lists one directory, and a `ListR` interface with does a recursive list if the remote is capable of it. This takes a lot of complexity out of the remotes which is a good thing hopefully!  If you take a look at the S3 remote you'll see how it has changed.

**Comment by ncw** [comment]
FYI [Here are the S3 changes](https://github.com/ncw/rclone/commit/8a6a8b9623bea1f600ad15eec2dc74ed37e65110#diff-7db3cb93944d57b2ffc803281c906018) to the List and ListR interfaces

**Comment by Kael9527** [comment]
Thanks for you ack!

Answer for you questions 
 
        QingCloud provided some free services for new users, And if you are interested to more experience, I can also provide a limited time free account to you(this may require you to provide an email), But I'm not sure if you can register in the UK and use it.

	About commented test case, It is can passing after the QingStor interface updated(may be long...).

	I'm already tried to run "go test -v github.com/ncw/rclone/qingstor" and through.

	About compatibility, Is qingstor one-way compatible with the S3 interface; This is for S3 users can smoothly migrate to QingStor.


The `list` interface change very well I think.   I'd be interested to know more. Such as, how do I test it after making a change.

**Comment by ncw** [comment]
OK I'll have a go at registering with QingCloud and see what happens!

> About compatibility, Is qingstor one-way compatible with the S3 interface; This is for S3 users can smoothly migrate to QingStor.

I'm not sure I understand. Would the existing rclone s3 backend work with QingStor?  And if so why would you use this one?

> The list interface change very well I think. I'd be interested to know more. Such as, how do I test it after making a change.

regenerate the *_test.go with `cd fstest/fstests && go generate` and that should have all the tests you'll need in.

**Comment by Kael9527** [comment]
> regenerate the *_test.go with cd fstest/fstests && go generate and that should have all the tests you'll need in.

OK, I'm got it. After that I will submit a new commit.

> I'm not sure I understand. Would the existing rclone s3 backend work with QingStor? And if so why would you use this one?

Sorry, maybe I didn't explain it clearly before, That's only a certain degree of compatibility, not complete. And having a separate backend is a a better choice.

File: docs/content/qingstor.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+date: "2017-04-14"
+---
+
+<i class="fa fa-amazon"></i> Qingstor
+---------------------------------------
+
+Paths are specified as `remote:bucket` (or `remote:` for the `lsd`
+command.)  You may put subdirectories in too, eg `remote:bucket/path/to/dir`.
+
+Here is an example of making an s3 configuration.  First run
```
Comment (seuffert) [inline docs/content/qingstor.md]: i think it should say "Qingstor" here , instead of "s3" ?

Comment (Kael9527) [inline docs/content/qingstor.md]: Thanks for reviewed!

I'm will fix it.

**Comment by Kael9527** [comment]
Are you review it?

**Comment by ncw** [comment]
I think the code looks good - and you've done a good job keeping up with recent changes.

I'd like to test it now but I'm having trouble getting a test account.  I tried to sign up for one, but it wouldn't accept my UK phone number.

Can you help me with that?  All I would use it for is for running the rclone integration tests!

Thanks

Nick

**Comment by Kael9527** [comment]
I'm sorry for the late reply.  I will try to apply for a test account for you. Please wait.

**Comment by ncw** [comment]
@Kael9527 any news on the test account?  I'd like to get this into 1.38 if possible.

**Comment by Kael9527** [comment]
@ncw   I jsut send a account and passwd to you mailbox "nick@craig-wood.com"

**Comment by ncw** [comment]
I've merged this now - thank you very much for your contribution :-)

I got all the tests working with a bit of modification of the tests and a bit of modificaton of the qingstor remote.

Do you have any plans to add MD5SUM support?  I note that you use multipart upload for everything so no files every have MD5SUMs - perhaps small ones could be uploaded in one part.

Also modtime support would be nice - I don't know if qingstor can have metadata or not.

**Comment by Kael9527** [comment]
>Do you have any plans to add MD5SUM support? I note that you use multipart upload for everything so no files every have MD5SUMs - perhaps small ones could be uploaded in one part.

That's what I'm going to do next.

> Also modtime support would be nice - I don't know if qingstor can have metadata or not.

QingStor will support custom metadata about this month, Then I can add it.

**Pull request closed without merging**